### ISSUE TITLE
Stable branches

### ISSUE THREAD

**Issue by JohannesLorenz** [LMMS/lmms#4780]
Currently, we're not getting 1.2.0 finished. Looking at the open issues and the time we need to fix them (and the amount of new issues) I don't expect it for the next 3 or 4 months, and even then, 1.2.1 is announced. This has drawbacks:

* We're rejecting developer's contributions, even if they're tested well, arguing that stable-1.2 has feature freeze. Developers get disappointed because features they develop now will be released in maybe 2 years. Developers may even start forks that will never be merged :broken_heart: 
* We're prioritizing in a strange way. E.g., we're delaying stuff like CALF updates, which will fix multiple effects, or Lv2, which bring us up to 1200 plugins, because of minor issues in one single plugin (like the SF2 player issues).

A typical reply is "Stable must not contain any new bugs, so we need to solve them all now". I agree for security related bugs, i.e.:

* Bugs where clipping occurs that can damage speakers or the listeners
* Bugs that remove files from your hard disk (seldom!)
* Bugs that allow hackers to exploit savefiles to attack your PC (IIRC we never had that)

For all other bugs, I'd say, leave them open. **It's not that bad if a new release has some bugs**. Firefox makes new releases fixes so often, marking all older releases as vulnerable. Does that hurt?

Our definition of "stable" is questionable. Somehow, the 1.2 features/fixes all got stable (by testing them for a long time). We could test 1.3 features the same way, but they'd still be rejected for 1.2, just because of "feature freeze".

**Conclusion:**

We need to release regularly and give new features a chance to get stable. We need a new definition of stable, e.g. let's say an issue that has been announced for stable tests and had no issues for 3 months is considered stable. Make new releases every 3 months, discarding all bugs (except for security bugs like the ones listed above).

We could even release every month. Again, look at Firefox...

**Poll:**

* :eyes: Don't release anything until all 1.2 issues are solved; Don't allow features from master to become stable for 1.2.
* :+1: Release on a regular base (e.g. every 3 months), even with open issues. Stable branches are abandoned every release.
* :tada: Other option, please specify in a comment

**Comment by zonkmachine** [comment]
> and even then, 1.2.1 is announced

My intention with opening the 1.2.1 milestone was to get some lesser important issues out of the way for the 1.2.0 release to speed things up. That's all. If we don't want that later, then we'll remove it.


> I don't expect it for the next 3 or 4 months,

I was rather hoping to see this released within a month.

**Comment by justnope** [comment]
On discord I've read one user recommend to another user to "install the unstable version because it is more stable than the stable version"
It's not a literal quote, but close.

Maybe a way to settle this is to compare the download statistics from lmms.io. How many downloads for 1.1.3 and how many for 1.2.0-rc7? The former should outnumber the latter by a wide margin. If not maybe we should consider releasing despite the bugs we have.

I'm not advocating to release right now. All I'm saying is if users start to ignore the beta label and use it like it was a released product, then what are we doing pretending it's still an RC?

**Comment by JohannesLorenz** [comment]
I think releasing RCs as full versions does not suffice, but we also need to give new features a chance to "also become stable". This means

1. either have stable-1.2, stable-1.3, stable-1.3.1, stable-1.3.2 etc in parallel to let any of them get stable separately. Sounds a bit like a branch hell, but maybe it will work well.
2. have only one stable branch at a time, merge it back after the release and then start the next one.

Aside from these two options (and the current option), does anyone have another idea? Let's collect ideas and then make a vote?

**Comment by justnope** [comment]
I'm in favor for having a separate stable branch. However it's implemented (per release version or global).
I think if we want to go that route it would be handy to have a portable version of lmms we can build. (definitely not for the next release, but maybe we can put it in master)

It should make things easier for alpha and beta testers to download and test the "master" branch. If they can't find anything wrong, we can promote it to the stable branch. I'm surprised at the number of people in discord who are willing to do tests.

In any case, if we want a stable branch, I don't think it's a developers only decision and with minor adjustments I think we can get enough testers on board so we can know when a feature can be upgraded to stable.

**Comment by musikBear** [comment]
One stable branch.

**Comment by Gabrielxd195** [comment]
Guys, I think the way the LMMS is distributed now is "annoying", because most novice users always download 1.1.3 because it's more affordable; however, that version 1.1.3 contains more errors than the current version 1.2 rc7. From rc4 onwards, the trial version exceeded 1.1.3, correcting one of the most serious errors: the use of vst and others, and functions that have worked well so far have been added. that is, for some time 1.2.0 should have been there as the official one.

Now my opinion as a newbie and user of lmms is that, in terms of priority to which we refer, we must order the characteristics and errors by nature of urgency, that is, we have the two versions: stable and test. I suggest that periodic updates be made in the stable version where '' small '' corrections are made such as security flaws, minor gui and code errors, etc. and add small basic functions such as: group removed with the "delete" key, cutting tool, etc. characteristics and "small" errors but that are urgent; while in the test versions they are done in random time, where you can add all the "big" features, such as: audio recording, ghost notes, etc. and correct larger errors that take longer to develop. functions and errors that are necessary for the development of lmms but that can wait longer, but once they are individually stabilized and tested they are added to the stable version. This way although the process is slower but it will be more progressive, because we can accept the contributions of the developers without having to postpone for another future version, correct errors in advance, we give more opportunity to new features, correct urgent character errors , we serve the needs of ordinary users, we serve the stable version. Well, it's just my humble opinion, but even if this idea is not valid, we have to find another better way to distribute it to lmms. regards

**Comment by BaraMGB** [comment]
There will never be a perfect time for releasing. Just release 1.2

**Comment by JohannesLorenz** [comment]
I turned this into a poll. If you haven't voted yet, please do now.

**Comment by zonkmachine** [comment]
I don't see that :+1: and :eyes: are mutually exclusive. There has been talks on changing the releases going forward but I basically missed that since I've been away for .long periods. As for :+1: , features from master aren't going into stable-1.2 no. Maybe we should have kept the door open longer as it's taken such a long time to complete this, at least I think so, but now it's really feature freeze and 1.2 is going to be released pretty soon.

**Comment by JohannesLorenz** [comment]
I'm closing this. The poll does not seem representative enough anymore.

**Comment by dvzrv** [comment]
> The poll does not seem representative enough anymore.

Not sure what that means, but I can also only recommend to release now (from a packager point of view).
This will ease confusion and streamline things a little again.
Also: One stable branch. I don't think there are enough developers to keep a multibranch thing alive.

Best of luck!
